charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_info.py

```diff
@@ -50,15 +88,41 @@ def some_event_function():
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 1
+LIBPATCH = 2
 
-
-RELATION_NAME = "gateway-info"
-INTERFACE_NAME = "istio-gateway-info"
+# Default relation and interface names. If changed, consistency must be kept
+# across the provider and requirer.
+DEFAULT_RELATION_NAME = "gateway-info"
+DEFAULT_INTERFACE_NAME = "istio-gateway-info"
 
 logger = logging.getLogger(__name__)
 
 
+def relation_preflight_checks(
+    model: Model, relation_name: str, requirer: bool = False
+) -> None:
+    """Series of checks for ensuring the relation is properly established.
+    Args:
+        model (Model): the juju model as seen by the provider/requirer unit
+        relation_name (str): the name of the relation
+        requirer (bool, optional): set to True if the check is run in the requirer class
+    Raises:
+        GatewayRelationMissingError when there is no relation between two units
+        GatewayRelationDataMissingError when there is no data in the data bag
+    """
+    # Do nothing if the unit is not leader
+    if not model.unit.is_leader():
+        logger.info("This unit is not leader, no action will be taken.")
+        raise RuntimeError("Unit is not leader")
```

Comment:
This feels like something the charm should be responsible for, not the library.  The charm would know if leadership matters for whether it uses this relation's data, we shouldn't impose that on the charm.

For example, if a user of this check offers HA, this means the non-leader units cannot check if they have relation data

Reply:
ah, I see the old one applied this check as well. I'm on the fence on whether we should immediately remove or treat it as something new for later

Reply:
overall check for leader is fine. The question if we should raise or not
https://github.com/canonical/prometheus-k8s-operator/blob/9d386da8023f91049252adab5e1db5fb9845251e/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1870

Reply:
I think I resolved this one by removing the check. Please review the updated version of the code.